[FEAT/#9] 도메인 모델 구현

hyunw9 opened the pull request:
## Related Issue 🚀
- closed #9 

## Work Description ✏️
- 도메인 모델을 POJO로 구현합니다. 

논의헀던 사항대로, User 객체 안에 여러 멤버들을 포함하는 구조로 작성했습니다. 
각 멤버들은 모두 `불변객체`로서, 수정이 일어날 경우 필드를 직접 수정하는 것이 아닌, 수정된 객체를 반환하는 형식으로 구현하고자 합니다. 

또한 프로젝트에서 중요하다고 생각되는 도메인 모델이기 때문에,, 가볍게 테스트 주도 개발을 한번 해보았습니다. 이렇게 하는게 맞는지..



## PR Point 📸

#### 다음은 구현 단계에서 제가 개인적으로 생각한 제약조건 및 검증입니다. 틀린 부분이 있을 수 있습니다. 

## Activity
https://github.com/sopt-makers/sopt-auth-backend/blob/77ef5e0472a4fd2f571c9a0ae32d806f6c23984d/domain/src/main/java/sopt/makers/authentication/user/Activity.java#L33-L40

TEAM 은 없을 수 있지만, PART, ROLE, Generation이 Null일 수 없기 때문에, 검증로직을 추가했습니다. 

## ActivityList
https://github.com/sopt-makers/sopt-auth-backend/blob/77ef5e0472a4fd2f571c9a0ae32d806f6c23984d/domain/src/main/java/sopt/makers/authentication/user/ActivityList.java#L32-L40

유저는 여러 활동을 할 수 있지만, 중복된 활동 정보는 존재할 수 없기 때문에 중복 검사를 진행합니다. 
O(N) 탐색입니다. 

https://github.com/sopt-makers/sopt-auth-backend/blob/77ef5e0472a4fd2f571c9a0ae32d806f6c23984d/domain/src/main/java/sopt/makers/authentication/user/ActivityList.java#L42-L47

Activity에서는 활동 내용들을 검증한다면, ActivityList는 Activity 자체가 Null인지 검증합니다.
당연히 Input으로 들어오는 Activity는 Null일 수 없기 때문에, `IllegalArgExption`을 반환합니다.

## Auth Platform
https://github.com/sopt-makers/sopt-auth-backend/blob/77ef5e0472a4fd2f571c9a0ae32d806f6c23984d/domain/src/main/java/sopt/makers/authentication/user/AuthPlatform.java#L5-L15

사용자는 Enum에 직접접근이 어렵기 때문에, String 형을 입력으로 받아 Enum을 반환하는 로직을 작성했습니다. 
e.g. client ("GOOGLE") 요청 -> server는 Enum.Google 반환 

위 방식은 검색이 필요한 Enum에 대부분 적용했습니다. 

## SocialAccount 
https://github.com/sopt-makers/sopt-auth-backend/blob/77ef5e0472a4fd2f571c9a0ae32d806f6c23984d/domain/src/main/java/sopt/makers/authentication/user/SocialAccount.java#L22-L26

SocialAccount 검색을 위해 들어오는 문자열이 Null이거나 비어있을 경우, 예외를 반환합니다. 

--- 

현재 반환하는 IllegalArgumentException은, 추후에 다른 예외로 전환이 가능합니다. 
리뷰 부탁드립니다.. 굽신굽신

hyunw9 committed:
[FEAT] Activity 구현

활동 정보를 담당하는 Activity입니다.
AcitivityList 일급객체를 두어, 중복 검증 등 로직을 처리하고자 하였습니다.
나머지 필드들은 ENUM으로 설정하였습니다.

hyunw9 committed:
[Feat] SocialAccount 구현

유저의 Oauth2 인증 정보를 담는 모델입니다. 플랫폼은 ENUM으로 설정하였습니다.

hyunw9 committed:
[FEAT] Profile 구현

유저의 개인 정보를 담당하는 Profile 모델입니다.

hyunw9 committed:
[FEAT] User 구현

아래 요소들을 가지고 있는 User모델입니다.
아직 구체적인 로직은 추가하지 않았지만, 생각했을 때 가장 기본적인 로직인 활동 이력 추가 (join) 만 추가헀습니다.

sung-silver commented on domain/src/main/java/sopt/makers/authentication/user/Activity.java:
```diff
@@ -0,0 +1,61 @@
+				|| activity.generation == null
+				|| activity.part == null
+				|| activity.role == null) {
+			throw new IllegalArgumentException("활동 정보가 비어있습니다..");
```
메세지에 .(마침표) 하나만 남겨서 통일성 있게 가져가면 좋을 것 같아요!

그리고 하나 더 궁금한게 저희 서비스 CustomException은 사용 안하는걸까요?? 예외마다 저렇게 텍스트로 예외메시지를 적으면 유지보수할 때 힘들 것 같아서요! 아직 논의를 안해서 이렇게 처리하신거라면 논의해보면 좋을 것 같습니다!

hyunw9 replied:
CustomException 사용하는거 선호합니다! 다만 저희가 따로 논의한적이 없기 때문에, 일단은 일률적으로 설정해놓았어요!

동규님이 공통모쥴 세팅하고 가볍게(?) 논의 해보면 좋을 것 같습니다!

yummygyudon replied:
저도 개인적으로 계층적으로 Custom Exception 을 관리하는 것을 좋아하는데요!!
이 부분도 빠르게 논의할 필요가 있어보이네요!
(금주 목요일 회의록에 쇼쇼쇽...)

<br/>

싱글모듈이 아닌 멀티모듈이기 때문에 
저희가 설계 시점부터 특정 모듈에 과도한 역할 수행과 이로 인한 비정상적인 의존관계 형성을 경계하고 지양하고자 했잖아요!!

이에 Exception class 와 Error/Success enum을 "**각 모듈별로 독립적인 체계를 구축해서 관리하는 것**"을 선호하실지 "**`support` 모듈에서 일괄적으로 관리할 것**"을 선호하실지 궁금합니다.

`api` 모듈에서는 `domain` & `support` 모듈 모두 주입 받는 구조이다 보니 ExceptionHandler & ControllerAdvice 설정에는 큰 문제가 없을 것으로 예상됩니다!!
그래서 가볍게 저희의 그라운드 룰만 정하면 될 것 같네요!!

yummygyudon replied:
`P2`

임원진 중 회장단의 경우, Part가 Null 일 수 있습니다!!
(ex. 다른 파트 참여 없이 회장단 활동만 가능)

hyunw9 replied:
저는 개인적으로 Domain에서 발생하는 Exception은 따로 관리하는게 어떨까 싶습니다. 

Domain에서 발생하는 예외 -> 비즈니스 예외
외 api, support, infra 예외 -> 기타 예외 

라고 생각하는데요, Support에 공통 Exception을 만들고, ExceptionAdvice를 달아버리면 `Domain` 모듈에서도 `Support`에 종속성이 생기니, 비즈니스 예외로 명명할 수 있는 예외는 도메인 모듈에서 관리하는게 좋을 것 같다고 생각해요!

yummygyudon replied:
> Domain에서 발생하는 예외 -> 비즈니스 예외
외 api, support, infra 예외 -> 기타 예외

위 개념에 대해 전적으로 동의합니다.
그렇다면 각 모듈마다 발생할 수 있는 에러에 대해서 각자 로깅할 수 있도록 해야겠네요..!!

logback은 api 모듈에만 설정한 후, 도메인에 정의했던 Exception를 Exception Handler에 등록하는 방식으로 생각하면 될까요?!

sung-silver commented on domain/src/test/java/sopt/makers/authentication/user/ActivityTest.java:
```diff
@@ -0,0 +1,126 @@
+		Activity activity = null;
+
+		// when & then
+		assertThatThrownBy(
+						() -> {
+							activityList.addActivity(activity);
+						})
+				.isInstanceOf(IllegalArgumentException.class);
```
여기 포매팅 적용된거 맞을까요??

hyunw9 replied:
spotless apply 하긴 했는데.. 조금 이상해 보이긴 합니다.

sung-silver commented on domain/src/test/java/sopt/makers/authentication/user/ActivityTest.java:
```diff
@@ -0,0 +1,126 @@
+	}
+
+	@Test
+	public void 같은_활동_이력은_동시에_존재할_수_없다() {
```
`@Display` 대신에 함수 이름을 한글로 작성하는걸 선호하시는건지 궁금합니다!

hyunw9 replied:
이러나 저러나 상관 없다는 주의입니다!

sung-silver replied:
이거도 저희끼리 맞추면 좋을 것 같네요!!~

yummygyudon replied:
저는 개인적으로 IDE 하이라이트가 신경쓰여서...ㅋㅋㅋㅋ
`@Display` 사용 +  영어 메서드 명명 방식을 선호하긴 합니다!!

hyunw9 replied:
`@Display` 는 동의하지만, 제가 영알못이라 영어로 작성하면 한눈에 안읽힐 때가 있더라구요.. 
테스트 명은 한글 어떠신지.. ㅎㅎ

yummygyudon replied:
쫗아욧!! 대신 메서드 명 포맷만 통일시키면 더 보기 좋을듯

sung-silver reviewed (approved):
저는 궁금한거 위주로 남긴 것 같습니다 수고하셨습니다!

yummygyudon commented on domain/src/main/java/sopt/makers/authentication/user/ActivityList.java:
```diff
@@ -0,0 +1,48 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class ActivityList {
```
`P3`

개인적으로 도메인 VO의 **객체명으로 자료구조 이름**이 포함되는 것을 지양합니다만
여러분들의 의견이 궁금합니다. (cc. @hyunw9 , @sung-silver )

물론 객체 명이 복수형인 것도 바람직하다고 생각하진 않기 때문에 (ex. `Activities.class`)
`User` 클래스 내부에서 `List<Activity>` 를 필드로 선언하고 List에 대한 검증을 진행하는 것은 어떨지 의견이 궁금합니다!

hyunw9 replied:
이 부분은 처음 구현 당시에, 한 유저가 여러 2번 이상의 활동 정보를 가질 수 있고, 정보 조회에 최근 기수를 받아오는 등의 로직이 들어갈 것 같아서, 따로 일급객체로 분리한 상태이긴 합니다. 
이런 부분들을 고려했어요. 

1. 가장 최근 활동 정보를 가져온다
2. 모든 활동 정보를 가져온다
3. 가장 마지막 활동 정보를 가져온다 

근데 해당 부분을 User에서 처리하면, User에 책임이 조금 분산될 수 있지 않을까 생각했습니다 !

yummygyudon replied:
그렇게 되면 
이후 infra 모듈을 통해 DB로 부터 조회해온 단일 유저에 대한 List Activity Entity객체에 대해
매번 하나 먼저 넣어서 생성하고 for문으로 `addActivity`를 수행하여 도메인 객체를 생성해줘야 하는 상황이 발생하는 것이 좋은 구조일지가 걱정이었습니다.

VO 객체라고 정의하기엔 가변적인 측면이 꽤 많은 도메인인 듯해서요!!

차라리 그렇다면 생성할 때, 생성자로 List<Activity> 를 주입 받고 `this.activities = ...`에 초기화 시키는 방식으로 적용하고
이후 Activity가 추가되는 경우 호출되는 `addActivity()`에서는 
```java
public ActivityList addActivity(Activity activity) {
    validateActivityDuplication(activity);
    validateActivityEmpty(activity);
    List<Activity> updatedActivities = new ArrayList<>();
    updatedActivities.addAll(this.activities);
    updatedActivities.add(activity);
    return new ActivityList(updatedActivities);
}
```
위와 같은 방식으로 구현하는 것이 불변객체로서의 도메인 POJO 의의가 지켜진다고 생각합니다.

sung-silver replied:
저도 Activities 좋은 것 같습니다!

yummygyudon commented on domain/src/main/java/sopt/makers/authentication/user/User.java:
```diff
@@ -0,0 +1,40 @@
+
+	private final Profile profile;
+
+	private final ActivityList activityHistory;
```
`P3`

복수의 `Activity`를 나타내는 필드인만큼 `acitivityHistory` 보다 `activities` 어떨지 제안드려봅니다!

hyunw9 replied:
좋습니다 👍👍

yummygyudon commented on domain/src/main/java/sopt/makers/authentication/user/User.java:
```diff
@@ -0,0 +1,40 @@
+
+	private final ActivityList activityHistory;
+
+	public User(Long id, SocialAccount socialAccount, Profile profile, Activity activity) {
+		this.id = id;
+		this.socialAccount = socialAccount;
+		this.profile = profile;
+		this.activityHistory = new ActivityList();
+		this.activityHistory.addActivity(activity);
+	}
```
`P2`

위와 같이 `List<Activity>`를 주입하지 않고 단일 `Activity` 객체만 주입하여 생성하는 방식으로 사용할 경우, 

여러 번의 활동이력이 존재하는 회원은 User 클래스를 생성하기 위해선 여러 Activity 중 랜덤하게 단일 `Activity`로 생성한 후
나머지 Activity를 `join` 메서드를 통해 주입해야하는 구조일까요?!!

hyunw9 replied:
구현해놓은 방식대로라면 YES 입니다. 

제가 생각한 사고의 방식은  `유저가 생성되는 시점` -> `회원가입` 이라고 생각했습니다.

즉, 유저는 각 플랫폼으로 회원가입 할 때만 최초로 생성되는 상황인거죠. 

활동 기수 정보가 없는 회원은 존재할 수 없기 때문에 

**모든 회원은 회원 가입 시 활동 정보를 기입한다** 라는 전제로 작성한 로직이긴 합니다. 

회원가입 외의 유저가 생성되는 케이스가 존재한다면 수정될 수 있을 것 같습니다.

yummygyudon commented on domain/src/main/java/sopt/makers/authentication/user/Profile.java:
```diff
@@ -0,0 +1,58 @@
+		return Objects.hash(name, email, phone, birthday);
+	}
+
+	public Profile updateName(String name) {
+		return new Profile(name, email, phone, birthday);
+	}
+
+	public Profile updateEmail(String email) {
+		return new Profile(name, email, phone, birthday);
+	}
+
+	public Profile updatePhone(String phone) {
+		return new Profile(name, email, phone, birthday);
+	}
+
+	public Profile updateBirthday(LocalDate birthday) {
+		return new Profile(name, email, phone, birthday);
+	}
```
`P1`

각 멤버들은 모두 불변객체로서, 수정이 일어날 경우 필드를 직접 수정하는 것이 아닌, 수정된 객체를 반환하는 형식으로 구현하고자 했는데
해당 `Profile` 객체만 상태값 변경 메서드를 정의한 이유가 따로 있을까요?!
(다른 VO 객체들은 별다른 변경 메서드가 없는 것 같아서 궁금했습니당ㅋㅋㅋ)

hyunw9 replied:
딱히 의미는 없습니다 ! 

다만, profile의 경우 유저에 의한 수정이 많이 일어날 수 있는 부분인 것 같아서, 이 부분만 일단 구현해 놓았습니다. ㅎ ㅎ

hyunw9 replied:
다른 VO도 변경이 필요하다면 얼마든지 추가할 수 있구요 !

yummygyudon reviewed (changes requested):
많이 늦어졌는데 몇 가지 의견과 궁금증을 달아보았습니다!!
편하게 봐주세요!!

---
전체적으로 Test를 진행할 때, `@ParameterizedTest` 와 같은 기능을 활용하여
given을 파라미터로 주입하는 방식을 활용하는 것도 좋아보입니다!!

hyunw9 reviewed (commented):
일단 당장 답변 가능 한 것만 달아보았습니다 ! 

나머지는 수정하면서 작성해보겠습니다 ㅎ

hyunw9 committed:
[FIX] 모든 테스트 @Display 적용

hyunw9 committed:
[FEAT] 멤버 이름 변경: activityHistory -> activities

yummygyudon commented on domain/src/test/java/sopt/makers/authentication/user/ActivityTest.java:
```diff
@@ -0,0 +1,135 @@
+		activityList.addActivity(activity1);
+
+		// when
+		Part part = Part.ANDROID;
```
이 `Part part = Part.ANDROID`는 어디에 쓰는 친구인고...?!

yummygyudon reviewed (changes requested):
한 번 더 리뷰를 진행해 보았는데요!!
VO로서 사용되며 내부 필드들의 "불변성 보장"이 필요한 class는 record 타입으로 구현하는 것이
타입으로서 불변성을 강제하고 가독성을 개선할 수 있을 것 같습니다!!

또한 도메인 객체 모두를 너무 불변성을 보장하려다 자칫 유연성이 저해될까 걱정이 되기도 하네요!!
몇 몇 필드들은 가변성이 짙은 요소들인 것 같아서요!!

여러분들의 의견 편하게 남겨주세요!! (@sung-silver , @hyunw9 )

hyunw9 committed:
[FEAT] 임원진 비즈니스 로직 검증 구현

test에 추가로 작성해 두었습니다.

hyunw9 committed:
[FEAT] 활동 추가 로직 수정

 List 형을 매개 변수로 받는 생성자를 추가했습니다.
그후, 새로운 활동 추가 이전에, 기존 활동을 복사하고, 새 활동을 추가해 반환하도록 구현했습니다.

hyunw9 committed:
[FEAT] VO Class -> Record 변경

단순 값 조회를 위해 사용되는 객체들을 Record로 변경하고, 검증 로직을 기본 생성자에서 수행하도록 변경했습니다.

hyunw9 committed:
[FEAT] ActivityList final -> final X 로 변경

기존 로직이 변경됨에 따라, ActivityList를 바꿔줘야 하기 때문에 Final을 해제했습니다.

sung-silver commented on domain/src/main/java/sopt/makers/authentication/user/User.java:
```diff
@@ -8,14 +8,13 @@ public class User {
+	private ActivityList activities;
 
-	public User(Long id, SocialAccount socialAccount, Profile profile, Activity activity) {
+	public User(Long id, SocialAccount socialAccount, Profile profile) {
```
저는 파라미터로 넘겨지는 값들에 해서는 불변성 보장을 위해 final 키워드를 붙이는편인데요 생성자에도 적용하는건 어떨까요?

sung-silver commented on domain/src/test/java/sopt/makers/authentication/user/UserTest.java:
```diff
@@ -63,31 +63,34 @@ public class UserTest {
 
 		// when
-		Activity activity1 = new Activity(34, null, Part.ANDROID, Role.MEMBER);
+		Activity activity1 = new Activity(35, null, Part.ANDROID, Role.MEMBER);
```
activity1보다 activityTerm34 (제가 쓰고도 좀 별론데.. 지금은 좋은 변수명이 생각이 안나요.. ㅜ) 등으로 변수가 직관적이면 좋을 것 같습니다!

yummygyudon replied:
저는 보동 테스트에서 비교 데이터에 대해서는
`given~`/`expected~` 혹은 `asIs~`/`toBe~` 등을 Prefix로 두고 하는 것을 선호하긴 합니다!!

sung-silver reviewed (commented):
저도 많이 배워가는 리뷰들이 있었네요! 개인적인 의견도 남겨보았습니다~

yummygyudon commented:
궁금한 점 몇 가지 남깁니다!!
(한 번에 꼼꼼히 드리지 못하고 나눠서 여러 번 드리게 된 점 양해 부탁드립니다...!!)

<br/>

1. **내부 field에 Wrapper 클래스를 사용하는 이유**

몇 몇 VO 내부에서 기본형 타입인 `int`나 `long`이 아닌 `Integer`와 `Long`을 사용하는 걸 확인했습니다.
null 처리하는 것에 있어 이점이 있어 Entity 등에서 요긴하게 쓰이는 타입이지만
기본적으로 저장 용량 낭비 및 연산하는 데에 있어 어려움(unboxing 후에 가능)이 존재한다는 특징이 있습니다.

불변 객체이며 VO이기 때문에 해당하는 대부분의 필드들은 not null이 보장되어야 할텐데
기본형 타입이 아닌 Wrapper 클래스를 사용하신 이유가 궁금해요!

<br/>

2.  **Test 시, given 용도의 Dummy Data를 ParameterizedTest로 수행하지 않고 직접 객체를 만든 이유**

모든 테스트 메서드 내에서 given 값에 대해 구성 요소들을 순차적으로 생성하는 로직이 작성되어 있는 것 같아요!

물론 객체 생성 프로세스를 명시하는 부분에서 좋을 순 있겠지만 테스트의 목적은 Dummy Data를 만드는 과정을 검증하는 것이 아닌
비교군과 기대군 간의 검증이 목적이기 때문에 자칫 가독성과 반복적인 코드 중복으로 자원 낭비가 될 . 수있다고 생각합니다.

또한 여러 메서드에서 반복되어 사용하는 데이터라면 한 번만 정의해서 재사용할 수있다는 장점도 가져갈 수 있을 것같습니다!!

여러가지 JUnit에서 지원하는 기능을 맘껏 활용해보는 것은 어떨까요?!

<br/>

---
우선 요종도까지인 것 같습니다!!
감사합니다! 😊

hyunw9 committed:
[FEAT] Wrapper -> Primitive 형 수정

hyunw9 commented:
> 또한 도메인 객체 모두를 너무 불변성을 보장하려다 자칫 유연성이 저해될까 걱정이 되기도 하네요!! 몇 몇 필드들은 가변성이 짙은 요소들인 것 같아서요!!

근데 저도 이 부분 동의합니다. 유연함을 가져가는게 좋을 것 같아요
